### DISCUSSION TITLE
Spring cache with async caffeine

### DISCUSSION THREAD

**Discussion by namila007** [ben-manes/caffeine#518] [Q&A] [answered]
Hi,

I'm currently using Caffeine with Spring Caffeine cache support implementation. I want to know is there any support for Async LoadingCache support on the spring cache? or if I want async caching do I want to move out with AsyncLoadingCache without spring caffeine cache support;
As I'm using spring-web flux, is this safe to use in a multi-threading environment?

**Comment by ben-manes** [comment, marked as answer]
My limited understanding is that Spring Cache does not support asynchronous caching, e.g. by using Reactor. The interfaces assume blocking calls where the values are opaque objects. You would have to ask the Spring folks for more advice, but I believe they have previously expressed no interest in asynchronous support. Unfortunately my experience with that framework is over a decade old, so I can't offer much help.

In the past, the Spring Cache authors have said their abstraction is only intended for simple, common cases. Any advanced cases should be handled by using the native apis. You should be able to use `AsyncLoadingCache` directly and convert between futures and monos, for use by spring-web.

**Reply by namila007** [reply]
yeh, recently I founded that the spring cache won't support Async as most of the caches also blocking [ISSUE LINK](https://github.com/spring-projects/spring-framework/issues/25173). I thought of moving to Redis as spring has reactive support to it. finally, I want to know that is this AsyncCache.. is thread-safe?  Thanks for the fast reply and to the lib :-)

**Reply by ben-manes** [reply]
Yep, `AsyncCache` is thread-safe.

**Reply by namila007** [reply]
nice, I will use AsyncCache for now (with Reactor support). thank you, Ben :)

**Comment by s-volkov-1** [comment]
Hello! Can I assume that plain (or manual) cache is non-blocking? When I compute values on my side and just want to use `cache.get(key) / cache.put(key, value)`?

**Reply by ben-manes** [reply]
I believe Spring Cache by default is non-blocking by using `cache.getIfPresent(key)` and `cache.put(key)`. If you use `sync = true` in its cache annotation then it will use a blocking `cache.get(key, mappingFunction)`.

You can do the same in your code. It is generally preferable to compute through the cache to avoid [stampedes](https://medium.com/@DoorDash/avoiding-cache-stampede-at-doordash-55bbf596d94b). That isn't always possible, e.g. recursive computations are not supported and have to be side loaded.
